Documentation/replacing-component.md

```diff
@@ -0,0 +1,266 @@
+# Proposal: Replacing `Component`
+
+
+- Author: Lukas Kollmer
+- Date: 2020-12-16
+- Draft Implementation: [experimental/dsl-node-protocols][branch_url]
+
+
+## Summary
+
+- We should replace `Component` with a collection of several, more fine-grained protocols, which match more closely to the different [kinds of components](https://github.com/Apodini/Apodini/tree/develop/Documentation/Communicational%20Patterns/3.%20Pattern%20Implementation) we eventually want to support
+- Instead of a single protocol, there should be a hierarchical structure of several protocols, similar to [how Swift models numeric types](https://files.lukaskollmer.de/embed/swift_numerics.png)
+- When defining a concrete component, the user would make that component conform to the protocol matching the component's intended use-case
+- We use the type system to ensure that components can only appear in places in the DSL where they actually make sense
+- For example, the leaves in the DSL's syntax tree must always be endpoints (ie components w/ a `handle()` function)
+- If possible, an ill-formed `WebService` shouldn't compile in the first place
```

Comment:
This summarizes the problems in a nice way!
As far as I have seen we so far would differentiate between the static structure of a web service (`Component`s/`EndpointProvidingNode`s) and dynamic behavior at the leafs of the `Component`/`EndpointProvidingNode` tree: `Handler`/`EndpointNode`. I would suspect that there is doing to be additional functionality from your side that will be needed to enable the communication between different components for dynamic deployments.